tower-http/src/services/fs/serve_dir.rs

```diff
@@ -285,26 +318,63 @@ async fn maybe_redirect_or_append_path(
     None
 }
 
-impl<ReqBody> Service<Request<ReqBody>> for ServeDir {
+impl<ReqBody, F, FResBody> Service<Request<ReqBody>> for ServeDir<F>
+where
+    F: Service<Request<ReqBody>, Response = Response<FResBody>> + Clone,
+    F::Error: Into<io::Error>,
+    F::Future: Send + 'static,
+    FResBody: http_body::Body<Data = Bytes> + Send + 'static,
+    FResBody::Error: Into<Box<dyn std::error::Error + Send + Sync>>,
+{
     type Response = Response<ResponseBody>;
     type Error = io::Error;
-    type Future = ResponseFuture;
+    type Future = ResponseFuture<ReqBody, F>;
 
     #[inline]
-    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        Poll::Ready(Ok(()))
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        if let Some(fallback) = &mut self.fallback {
+            fallback.poll_ready(cx).map_err(Into::into)
+        } else {
+            Poll::Ready(Ok(()))
+        }
     }
 
     fn call(&mut self, req: Request<ReqBody>) -> Self::Future {
-        let mut full_path = match self.variant.full_path(&self.base, req.uri().path()) {
+        // `ServeDir` doesn't care about the request body but the fallback might. So move out the
```

Comment:
All methods that are reasonable for asset routes don't have a body anyways, right? By the way, does `ServeDir` still serve files for non-GET requests? I think it did that at some point.

Reply:
> All methods that are reasonable for asset routes don't have a body anyways, right?

Yes. At least I can't think of a use case for serving a static file and having a request body.

> By the way, does ServeDir still serve files for non-GET requests? I think it did that at some point.

It still accepts all methods. Perhaps we should change that to respond with 405 for non-`GET` requests.

Reply:
There is proper handling of HEAD now, right? That should probably stay. Not sure about OPTIONS, probably okay to have it return 405 if the user doesn't have a cors middleware set up.